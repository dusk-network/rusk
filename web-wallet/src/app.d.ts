declare namespace svelteHTML {
  interface HTMLAttributes<T> {
    "on:wizardstepchange"?: (
      event: CustomEvent<{ step: number; stepsCount: number }>
    ) => void;
  }
}

/* Aliases for common w3sper's types */

type AccountBalance = import("@dusk/w3sper").AccountBalance;
type AccountSyncer = import("@dusk/w3sper").AccountSyncer;
type AddressBalance = import("@dusk/w3sper").AddressBalance;
type AddressSyncer = import("@dusk/w3sper").AddressSyncer;
type Bookmark = import("@dusk/w3sper").Bookmark;
type Gas = import("@dusk/w3sper").Gas;
type Network = import("@dusk/w3sper").Network;
type Profile = import("@dusk/w3sper").Profile;
type ProfileGenerator = import("@dusk/w3sper").ProfileGenerator;
type StakeInfo = import("@dusk/w3sper").StakeInfo;
