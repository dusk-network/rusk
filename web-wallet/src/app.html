<!doctype html>
<html lang="en">
  <head>
    <title>Dusk Web Wallet</title>
    <meta charset="utf-8" />
    <link rel="icon" type="image/png" href="%sveltekit.assets%/favicon.png" />
    <link
      rel="icon"
      type="image/svg+xml"
      href="%sveltekit.assets%/favicon.svg"
    />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    %sveltekit.head%
  </head>
  <body
    data-sveltekit-preload-code="viewport"
    data-sveltekit-preload-data="hover"
  >
    <div id="outer-container">%sveltekit.body%</div>
  </body>
</html>
