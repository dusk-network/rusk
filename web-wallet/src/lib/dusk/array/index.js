export { default as arraysEqual } from "./arraysEqual";
export { default as arrayMax } from "./arrayMax";
export { default as arrayMaxByKey } from "./arrayMaxByKey";
export { default as shuffleArray } from "./shuffleArray";
export { default as findFirstNMatches } from "./findFirstNMatches";
