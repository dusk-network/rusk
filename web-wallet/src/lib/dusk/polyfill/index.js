/** @overview Import the module to load all polyfills at once */

import "./asyncIterator";
import "./promiseWithResolvers";

// eslint-disable-next-line import/no-unresolved
import "web-streams-polyfill/polyfill";
