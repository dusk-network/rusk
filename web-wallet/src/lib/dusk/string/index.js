export { default as calculateAdaptiveCharCount } from "./calculateAdaptiveCharCount";
export { default as cleanNumberString } from "./cleanNumberString";
export { default as hexStringToBytes } from "./hexStringToBytes";
export { default as isValidEvmAddress } from "./isValidEvmAddress";
export { default as makeClassName } from "./makeClassName";
export { default as middleEllipsis } from "./middleEllipsis";
export { default as randomUUID } from "./randomUUID";
export { default as truncateDecimals } from "./truncateDecimals";
