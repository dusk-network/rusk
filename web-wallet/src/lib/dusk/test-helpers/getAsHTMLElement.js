/**
 * Utility function to force the type checker to see
 * the return value of `querySelector` as a HTMLElement.
 *
 * @param {HTMLElement} container
 * @param {String} selector
 * @returns {HTMLElement}
 */
const getAsHTMLElement = (container, selector) =>
  /** @type {HTMLElement} */ (container.querySelector(selector));

export default getAsHTMLElement;
