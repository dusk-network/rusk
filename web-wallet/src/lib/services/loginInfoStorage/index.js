import { compose, isNull, mapValuesWith, unless } from "lamb";

import { base64ToBytes, bytesToBase64 } from "$lib/dusk/base64";

const storeKey = `${CONFIG.LOCAL_STORAGE_APP_KEY}-login`;
const fromStorageString = unless(
  isNull,
  compose(mapValuesWith(base64ToBytes), JSON.parse)
);
const toStorageString = compose(JSON.stringify, mapValuesWith(bytesToBase64));

const loginInfoStorage = {
  /** @returns {WalletEncryptInfo | null} */
  get() {
    return fromStorageString(localStorage.getItem(storeKey));
  },

  remove() {
    localStorage.removeItem(storeKey);
  },

  /** @param {WalletEncryptInfo} info */
  set(info) {
    localStorage.setItem(storeKey, toStorageString(info));
  },
};

export default loginInfoStorage;
