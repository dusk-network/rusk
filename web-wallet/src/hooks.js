/** @see https://github.com/dusk-network/rusk/issues/3570 */
import "$lib/dusk/polyfill";
