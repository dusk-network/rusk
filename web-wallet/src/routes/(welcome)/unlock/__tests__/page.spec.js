import {
  afterAll,
  afterEach,
  beforeAll,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { generateMnemonic } from "bip39";
import { get } from "svelte/store";
import { ProfileGenerator } from "$lib/vendor/w3sper.js/src/mod";

import { getKey, setKey } from "lamb";

import { getAsHTMLElement } from "$lib/dusk/test-helpers";
import * as navigation from "$lib/navigation";
import { settingsStore, walletStore } from "$lib/stores";
import {
  encryptMnemonic,
  getSeedFromMnemonic,
  profileGeneratorFrom,
} from "$lib/wallet";
import loginInfoStorage from "$lib/services/loginInfoStorage";

import UnlockWallet from "../+page.svelte";

/** @param {HTMLElement} container */
function getTextInput(container) {
  // eslint-disable-next-line no-extra-parens
  return /** @type {HTMLInputElement} */ (
    container.querySelector("[type='password']")
  );
}

describe("Unlock Wallet", async () => {
  const mnemonic = generateMnemonic();
  const pwd = "some pwd";
  const loginInfo = await encryptMnemonic(mnemonic, pwd);
  const seed = getSeedFromMnemonic(mnemonic);
  const userId = await profileGeneratorFrom(seed)
    .then(getKey("default"))
    .then(getKey("address"))
    .then(String);

  const getErrorElement = () => document.querySelector(".banner--error");
  const gotoSpy = vi.spyOn(navigation, "goto");

  /**
   * Sometimes a "DatabaseClosedError: Database has been closed" is
   * thrown when running this test (never happened running it in isolation).
   *
   * As I can't pinpoint what's causing it (all connections are opened before
   * db operations), I added the mocked implementation as here we don't care
   * about running `init` for real.
   */
  const initSpy = vi.spyOn(walletStore, "init").mockResolvedValue(void 0);

  afterEach(async () => {
    cleanup();
    gotoSpy.mockClear();
    initSpy.mockClear();
    settingsStore.reset();
    walletStore.reset();
  });

  afterAll(async () => {
    gotoSpy.mockRestore();
    initSpy.mockRestore();
  });

  describe("Mnemonic phrase workflow", () => {
    it("should render the Unlock Wallet page and show the field to enter the mnemonic phrase, if there is no login info stored", () => {
      const { container } = render(UnlockWallet, {});

      expect(container.firstChild).toMatchSnapshot();
    });

    it("should show an error message if the user enters an invalid mnemonic", async () => {
      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      expect(getErrorElement()).toBeNull();

      await fireEvent.input(textInput, { target: { value: "foo bar" } });
      await fireEvent.submit(form, { currentTarget: form });

      const errorElement = await vi.waitUntil(getErrorElement);
      const selectedText = textInput.value.substring(
        Number(textInput.selectionStart),
        Number(textInput.selectionEnd)
      );

      expect(initSpy).not.toHaveBeenCalled();
      expect(errorElement?.textContent).toMatch("Invalid mnemonic phrase");
      expect(textInput).toHaveFocus();
      expect(selectedText).toBe(textInput.value);
    });

    it("should redirect to the Restore flow if the user inputs a valid mnemonic with no prior wallet created", async () => {
      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: mnemonic } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe("");
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    it("should redirect to the Restore flow the user inputs a valid mnemonic different from the last one used", async () => {
      const currentUserID = "some-user-id";
      settingsStore.update(setKey("userId", currentUserID));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: mnemonic } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe(currentUserID);
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    it("should unlock the Wallet if the entered mnemonic is the last one used", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: mnemonic } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(ProfileGenerator));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });

    it("should trim and lower case the entered mnemonic before validating it", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, {
        target: { value: `  \t${mnemonic.toUpperCase()} \t  ` },
      });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(ProfileGenerator));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });
  });

  describe("Password workflow", () => {
    beforeAll(() => {
      loginInfoStorage.set(loginInfo);

      return () => loginInfoStorage.remove();
    });

    it("should show the password field and the link to restore the wallet if there is login info stored", () => {
      const { container } = render(UnlockWallet, {});

      expect(container.firstChild).toMatchSnapshot();
    });

    it("should show an error message if the user enters the wrong password", async () => {
      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      expect(getErrorElement()).toBeNull();

      await fireEvent.input(textInput, { target: { value: "foo bar" } });
      await fireEvent.submit(form, { currentTarget: form });

      const errorElement = await vi.waitUntil(getErrorElement);
      const selectedText = textInput.value.substring(
        Number(textInput.selectionStart),
        Number(textInput.selectionEnd)
      );

      expect(initSpy).not.toHaveBeenCalled();
      expect(errorElement?.textContent).toMatch("Invalid password");
      expect(textInput).toHaveFocus();
      expect(selectedText).toBe(textInput.value);
    });

    /**
     * This is not a possible situation, in theory, but
     * the workflow is able to deal with it.
     */
    it("should redirect to the Restore flow if the user inputs the correct password with no prior wallet created", async () => {
      settingsStore.update(setKey("userId", ""));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });

      expect(getErrorElement()).toBeNull();

      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getErrorElement()).toBeNull();
      expect(get(settingsStore).userId).toBe("");
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    /**
     * This is not a possible situation, in theory, but
     * the workflow is able to deal with it.
     */
    it("should redirect to the Restore flow if the user inputs the correct password for a mnemonic different from the last one used", async () => {
      const currentUserID = "some-user-id";

      settingsStore.update(setKey("userId", currentUserID));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });

      expect(getErrorElement()).toBeNull();

      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    it("should unlock the Wallet is the entered password is for the last used mnemonic", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(ProfileGenerator));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });

    it("should trim the entered password before validating it", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(UnlockWallet, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, {
        target: { value: `  \t${pwd} \t  ` },
      });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(ProfileGenerator));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });
  });
});
