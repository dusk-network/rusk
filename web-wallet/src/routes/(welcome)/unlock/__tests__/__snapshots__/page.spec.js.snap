// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Unlock Wallet > Mnemonic phrase workflow > should render the Unlock Wallet page and show the field to enter the mnemonic phrase, if there is no login info stored 1`] = `
<section
  class="login svelte-1e8gc2s"
>
  <h2
    class="h1"
  >
    Unleash 
    <mark>
      RWA
    </mark>
     and
    <br />
     
    <mark>
      Decentralized Finance
    </mark>
  </h2>
   
  <div
    class="login svelte-1e8gc2s"
  >
    <article
      class="dusk-card dusk-card--gap-medium"
    >
      <div
        class="dusk-card__header-container"
      >
        <header
          class="dusk-card__header"
          slot="header"
        >
          <div
            class="dusk-card__heading-container"
          >
            <h3
              class="h4"
            >
              Mnemonic phrase
            </h3>
             
            <div
              class="dusk-card__heading-icons"
            >
              
              <svg
                class="dusk-icon dusk-icon--size--default"
                role="graphics-symbol"
                viewBox="0 0 24 24"
              >
                <path
                  d="M21 18H15V15H13.3C12.2 17.4 9.7 19 7 19C3.1 19 0 15.9 0 12S3.1 5 7 5C9.7 5 12.2 6.6 13.3 9H24V15H21V18M17 16H19V13H22V11H11.9L11.7 10.3C11 8.3 9.1 7 7 7C4.2 7 2 9.2 2 12S4.2 17 7 17C9.1 17 11 15.7 11.7 13.7L11.9 13H17V16M7 15C5.3 15 4 13.7 4 12S5.3 9 7 9 10 10.3 10 12 8.7 15 7 15M7 11C6.4 11 6 11.4 6 12S6.4 13 7 13 8 12.6 8 12 7.6 11 7 11Z"
                />
              </svg>
              
            </div>
          </div>
        </header>
      </div>
       
      <div
        class="dusk-card__body-container"
      >
        <form
          class="login__form svelte-1e8gc2s"
        >
          <input
            aria-label="Mnemonic phrase"
            autocomplete="current-password"
            autofocus=""
            class="dusk-textbox dusk-textbox-password"
            name="Mnemonic phrase"
            placeholder="Mnemonic phrase"
            required=""
            type="password"
          />
          
           
           
          <button
            class="dusk-button dusk-button--type--submit dusk-button--variant--primary dusk-button--size--default"
            type="submit"
          >
            <span
              class="dusk-button__text"
            >
              Unlock Wallet
            </span>
          </button>
           
        </form>
      </div>
       
    </article>
    
  </div>
   
  <footer
    class="login-footer svelte-1e8gc2s"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled"
      href="/some-base-path/setup"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
  </footer>
</section>
`;

exports[`Unlock Wallet > Password workflow > should show the password field and the link to restore the wallet if there is login info stored 1`] = `
<section
  class="login svelte-1e8gc2s"
>
  <h2
    class="h1"
  >
    Unleash 
    <mark>
      RWA
    </mark>
     and
    <br />
     
    <mark>
      Decentralized Finance
    </mark>
  </h2>
   
  <div
    class="login svelte-1e8gc2s"
  >
    <article
      class="dusk-card dusk-card--gap-medium"
    >
      <div
        class="dusk-card__header-container"
      >
        <header
          class="dusk-card__header"
          slot="header"
        >
          <div
            class="dusk-card__heading-container"
          >
            <h3
              class="h4"
            >
              Password
            </h3>
             
            <div
              class="dusk-card__heading-icons"
            >
              
              <svg
                class="dusk-icon dusk-icon--size--default"
                role="graphics-symbol"
                viewBox="0 0 24 24"
              >
                <path
                  d="M21 18H15V15H13.3C12.2 17.4 9.7 19 7 19C3.1 19 0 15.9 0 12S3.1 5 7 5C9.7 5 12.2 6.6 13.3 9H24V15H21V18M17 16H19V13H22V11H11.9L11.7 10.3C11 8.3 9.1 7 7 7C4.2 7 2 9.2 2 12S4.2 17 7 17C9.1 17 11 15.7 11.7 13.7L11.9 13H17V16M7 15C5.3 15 4 13.7 4 12S5.3 9 7 9 10 10.3 10 12 8.7 15 7 15M7 11C6.4 11 6 11.4 6 12S6.4 13 7 13 8 12.6 8 12 7.6 11 7 11Z"
                />
              </svg>
              
            </div>
          </div>
        </header>
      </div>
       
      <div
        class="dusk-card__body-container"
      >
        <form
          class="login__form svelte-1e8gc2s"
        >
          <input
            aria-label="Password"
            autocomplete="current-password"
            autofocus=""
            class="dusk-textbox dusk-textbox-password"
            name="Password"
            placeholder="Password"
            required=""
            type="password"
          />
          
           
           
          <button
            class="dusk-button dusk-button--type--submit dusk-button--variant--primary dusk-button--size--default"
            type="submit"
          >
            <span
              class="dusk-button__text"
            >
              Unlock Wallet
            </span>
          </button>
           
          <a
            aria-disabled="false"
            class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default"
            href="/some-base-path/setup/restore"
          >
            <span
              class="dusk-anchor-button__text"
            >
              Forgot Password?
            </span>
            
          </a>
        </form>
      </div>
       
    </article>
    
  </div>
   
  <footer
    class="login-footer svelte-1e8gc2s"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled"
      href="/some-base-path/setup"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
  </footer>
</section>
`;
